### PR TITLE
Replace rustup with running command

### LINKED ISSUE #806: Ctrl+C on `cargo run`
When I Ctrl+C on cargo run and my Rust program is set up to intercept the signal, the Rustup process stops and returns control to the terminal, even though my Rust program (and indeed Cargo) may still be running. I've observed this on the latest version of Rustup.

I think it would be best for Rustup to do something similar to what `cargo run` does, and intercept Ctrl+C when running a program, and ignore it – instead, it should just check regularly to see if the program (`cargo`) has exited.

Note: this is confirmed not to be a `cargo` issue, as per https://github.com/rust-lang/cargo/issues/3281.

### LINKED ISSUE #845: cargo: exit on signal swallowed by rustup wrapper
See https://github.com/rust-lang/cargo/issues/3367 - the difference:
```
$ cargo run
   Compiling rust-test v0.1.0 (file:///home/gbr/devel/rust-test)
    Finished debug [unoptimized + debuginfo] target(s) in 0.14 secs
     Running `target/debug/rust-test`
$ target/debug/rust-test
[1]    7789 segmentation fault  target/debug/rust-test
```
The exit on signal is hidden by rustup's cargo wrapper.

### PR DESCRIPTION
**Pull request by KalitaAlexey**
I copied the implementation from [Cargo](https://github.com/rust-lang/cargo/blob/master/src/cargo/util/process_builder.rs#L92).

I don't know if I can add any tests for that.

I checked on Ubuntu and Windows.

Fixes #806
Fixes #845

### CODE REVIEW

**Comment by KalitaAlexey** [comment]
@birkenfeld,
Can you check if it fixes your problem?

**Comment by KalitaAlexey** [comment]
@nrc,
Don't you know if @brson is on vocation?

**Comment by KalitaAlexey** [comment]
@brson,
I don't know why it failed.

File: src/rustup/command.rs [hunk 1]
```diff
@@ -135,24 +135,38 @@ fn telemetry_rustc<S: AsRef<OsStr>>(mut cmd: Command,
 fn run_command_for_dir_without_telemetry<S: AsRef<OsStr>>(
     mut cmd: Command, arg0: &str, args: &[S]) -> Result<()>
 {
-    cmd.args(&args);
+    #[cfg(unix)]
+    fn run(mut command: Command, arg0: &str) -> Result<()> {
+        use std::os::unix::process::CommandExt;
+        let error = command.exec();
+        Err(error).chain_err(|| rustup_utils::ErrorKind::RunningCommand {
```
Comment (nagisa) [inline src/rustup/command.rs]: You’ll probably want to chain the error after calling run, like this:

```
run(cmd, arg0).chain_err(|| ...)
```

instead. That greatly cuts down on the amount of code needed.

Comment (nagisa) [reply src/rustup/command.rs]: See for example similar code in [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz/blob/master/src/main.rs#L271)

Comment (KalitaAlexey) [reply src/rustup/command.rs]: Sounds reasonable. Thanks.

**Comment by brson** [comment]
Hi @KalitaAlexey.

Thanks for the fixes, and sorry for the delay.

This does solve the problem, but I want to solve these two issues a different way, and the main reason is that rustup wants to be able to perform work after the invoked subcommand is run (e.g. recording telemetry about the results). I am not prepared to give up that capability yet.

I'd instead prefer to do something more along the lines described in the op of https://github.com/rust-lang-nursery/rustup.rs/issues/806, installing a signal handler and capturing Ctrl+C. Likewise the signal-swallowing can be handled by examining how the subprocess terminated.

**Comment by KalitaAlexey** [comment]
@brson,
If you know how to do that, you should do it because I don't know.
I need to think about it.

**Comment by nagisa** [comment]
Theres another good reason to exec that is proper status reporting and
debugability. Namely currently the exit code is always 1 regardless of
rustc editting with code 101 or sogsegving. You cannot use gbd through the
wrapper that shadows the name of the rustc either which makes the use case
a pain. Exec solves all of these.

**Comment by birkenfeld** [comment]
I agree with @nagisa - especially in that rustup-rustc/rustup-cargo should behave the same as normal rustc/cargo. This is otherwise very confusing to users who shouldn't have to care about how their toolchain was installed.

**Comment by Diggsey** [comment]
I'll close this, since we're not planning on using exec - further discussion about whether this is the right decision can happen on the issue (#31)

**Pull request closed without merging**